3D Tiles - Support batch table binary

lilleyse opened the pull request:
For https://github.com/AnalyticalGraphicsInc/3d-tiles/issues/32

This adds support for a batch table binary section to b3dm, i3dm, and pnts.

To do:

* [x] Need `instancedWithBatchTableBinary` tileset for testing. (This causes one test to fail)
* [x] ~~Possibly remove `handleTranslucent` check for pnts, should be able to support transparent and opaque together.~~

pjcozzi commented:
In the Sandcastle example with the new batched point cloud, picking works well when zoomed out, but when zoomed in, it often flickers between the back and front faces of the sphere as the mouse moves, which is expected.

No action now, but when we do, for example, the edge-preserving blur, we'll need to determine how to "blur" the pick ids, e.g., think more like nearest texture filtering, not linear.

pjcozzi commented on Source/Core/ComponentDatatype.js:
```diff
@@ -297,5 +297,36 @@ define([
+     *
+     * @exception {DeveloperError} name is not a valid value.
+     */
+    ComponentDatatype.fromName = function(name) {
```
Add to CHANGES.md since this is a public class.

pjcozzi replied:
Could also cherry-pick this into master.

pjcozzi commented on Source/Scene/Batched3DModel3DTileContent.js:
```diff
@@ -199,25 +199,49 @@ define([
-        this._featuresLength = batchLength;
         byteOffset += sizeOfUint32;
 
+        // TODO : remove this legacy check before merging into master
```
@lilleyse when this is merged, can you also email the forum about this small breaking change for b3dm?

lilleyse replied:
Sure

pjcozzi commented on Source/Scene/Batched3DModel3DTileContent.js:
```diff
@@ -199,25 +199,49 @@ define([
+            batchLength = batchTableJsonByteLength;
+            batchTableJsonByteLength = batchTableBinaryByteLength;
+            batchTableBinaryByteLength = 0;
+            console.log('Warning: b3dm header is using the legacy format [batchLength] [batchTableByteLength]. Please update to the latest b3dm format: https://github.com/AnalyticalGraphicsInc/3d-tiles/blob/master/TileFormats/Batched3DModel/README.md.');
```
Add the new "[batchTableJsonByteLength] [batchTableBinaryByteLength] [batchLength]" to the string so it is more obvious what changed.

pjcozzi commented on Source/Scene/Batched3DModel3DTileContent.js:
```diff
@@ -199,25 +199,49 @@ define([
+            var batchTableBinary;
+            if (batchTableBinaryByteLength > 0) {
+                // Has a batch table binary
+                batchTableBinary = new Uint8Array(arrayBuffer, byteOffset, batchTableBinaryByteLength);
```
I don't know how the GC works, but I wonder if holding on to this subset of the original `arrayBuffer` forces the GC to keep the entire `arrayBuffer` in memory.  @shunter do you know?

pjcozzi commented on Source/Scene/Cesium3DTileBatchTableResources.js:
```diff
@@ -282,27 +297,116 @@ define([
     };
 
+    // TODO : put this helper somewhere else?
+    function getComponentCountFromType(type) {
```
Perhaps here and below are similar to [getModelAccessor.js](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Scene/getModelAccessor.js)?  That is private so it can be updated/renamed and even cherry-picked into master early.

pjcozzi commented on Source/Scene/Cesium3DTileBatchTableResources.js:
```diff
@@ -2,14 +2,19 @@
 define([
         '../Core/arrayFill',
         '../Core/Cartesian2',
+        '../Core/Cartesian3',
```
Should we rename:
* `Cesium3DTileBatchTableResources` -> `Cesium3DTileBatchTable`

I think I added "Resources" because of the batch table texture, but perhaps this rename would be more concise and mirror `Cesium3DFeatureTable`.

lilleyse replied:
`Cesium3DFeatureTable` is also `Cesium3DFeatureTableResources`. I agree that it's worth making both more concise.

pjcozzi commented on Source/Scene/Cesium3DTileBatchTableResources.js:
```diff
@@ -282,27 +297,116 @@ define([
+        }
+    }
+
+    Cesium3DTileBatchTableResources.prototype.setBatchTable = function(json, binary) {
```
Is it reasonable to replace this function with the constructor?  Or does that make the users of this class more difficult to implement?

lilleyse replied:
This should be easily doable.

pjcozzi commented on Source/Scene/Instanced3DModel3DTileContent.js:
```diff
@@ -272,15 +269,20 @@ define([
-            var batchTableString = getStringFromTypedArray(uint8Array, byteOffset, batchTableJSONByteLength);
-            batchTableResources.batchTable = JSON.parse(batchTableString);
-            byteOffset += batchTableJSONByteLength;
+        if (batchTableJsonByteLength > 0) {
```
Perhaps there should be a helper function for parsing this part of the header (and maybe for the feature table properties as well?).  Could be a static `from` function on `Cesium3DTileBatchTableResources` or the constructor.

lilleyse replied:
Personally I prefer it how it is since it keeps tile parsing all in the same place, and interacts easily with the `byteOffset` and other things.

pjcozzi replied:
OK, yeah, it is pretty coupled.

pjcozzi commented on Source/Scene/Points3DTileContent.js:
```diff
@@ -148,19 +155,42 @@ define([
         }
     });
 
+    function createFeatures(content) {
```
@lilleyse do you recall:

> The Cesium API should not expose each point as a feature; the overhead is too high

from https://github.com/AnalyticalGraphicsInc/3d-tiles/issues/22#issuecomment-231777777.

I suppose now is the time to think this through.

My thought here was that per-point `Cesium3DTileFeature` is just a incredible amount of memory overhead.  Instead, per-point batch table values would be stored in batch table textures and accessed in the shader for exploration and visualization (e.g., "show only points with some property > 0.5" or "map this property to this color ramp.").

The downside here is that the user can't inspect and change the per-point properties in JavaScript.  Perhaps the best option is to make this an option that defaults to `false`?

lilleyse replied:
So the terminology may be confusing since a single point is usually called a feature, but in this case the number of `Cesium3DTileFeature`s that are created is not per-point but instead per batch, so there would be 8 in the `PointsWithBatchTableBinary` example, rather than 1000.

But yeah, if users decide to define batch table values per-point, this will definitely be a memory concern and we will need to use an idea like you suggested.

pjcozzi replied:
> instead per batch

Do you mean batch id?

lilleyse replied:
Ah yeah, per batch id.

pjcozzi replied:
OK, I still think this has the potential to introduce huge overhead for naive tiles or tiles where points truly have a ton of unique values - probably true of many datasets that do not go through a cleaning pipeline.  How hard is it to add a flag as I suggested even if it defaults to `true`?

lilleyse replied:
Maybe a good enough check is seeing if `BATCH_LENGTH` equals `POINTS_LENGTH`, or something similar.

pjcozzi replied:
And then what?  Silently disabling this?  We can't do that.  Log a warning?  I don't think that is the right behavior.

lilleyse replied:
I think I have a better solution built off the discussion above:

If the BATCH_ID semantic is present, then the points are considered batched. This is the right technique for differentiating physical features within the point cloud, like doors vs windows vs walls. A `Cesium3DTileBatchTable` is used to individually select, color, and pick sections, as well as get/set batch properties.

If the BATCH_ID is not present, then the batch table stores per-point properties. In order to avoid the memory overhead of creating a `Cesium3DTileFeature`/color/show/pickID for every point, a `Cesium3DTileBatchTable` is not created. Instead the properties are inputted to the shader as vertex attributes (much easier to work with than textures and all styling happens in the vertex shader anyways) and styled in the shader. The properties cannot be get/set.

So overall the concept is, if BATCH_IDs are defined for the tile, then the point cloud is "batched" and uses a batch table like normal. Otherwise there is no batch table, and per-point properties are only available to the shader for styling.

pjcozzi replied:
Sounds good to me.  Is this already implemented?

Also, please add a discussion of this to the `pnts` spec so users know how to create per-point vs. per-feature metadata.  Keep the discussion generic, not tied to the Cesium implementation, which basically just does not support AGI-level get/set for per-point metadata.

pjcozzi replied:
Can you also add a comment here about this so readers don't assume a one-to-one relationship between points-and-features.

lilleyse replied:
Yes it's implemented in my latest changes. I'll make sure to include this in the spec and also add a comment in the code.

lilleyse replied:
Basically it's the difference between `PointsBatched` and `PointsWithPerPointProperties`.

pjcozzi commented on Source/Scene/Points3DTileContent.js:
```diff
@@ -367,6 +447,10 @@ define([
         }
 
+        if (hasBatchIds) {
+            vs += 'attribute float a_batchId; \n';
```
Did you verify this, for example, by temporarily tweaking the shader to map this to a color ramp?

lilleyse replied:
Just did, looks good:
![capture](https://cloud.githubusercontent.com/assets/915398/17933877/c127ed76-69e3-11e6-8dd2-69f01bce33ea.JPG)

pjcozzi replied:
Nice!

pjcozzi commented:
Can you verify that test coverage is solid?

Looks great!

lilleyse commented:
> In the Sandcastle example with the new batched point cloud, picking works well when zoomed out, but when zoomed in, it often flickers between the back and front faces of the sphere as the mouse moves, which is expected.

I should be able to make this better by using normals and doing back-face culling on the points.

lilleyse commented:
Ah..... I accidentally committed some large pnts tiles, I'll squash my changes with smaller versions.

lilleyse commented:
Updated.

This also lays the foundation for dynamic styling in the shader. Here is a point cloud that stores a `secondaryColor` (shades of red) and `temperature` (3d noise values) as per-point properties in the batch table, loaded in as vertex attributes and styled in the vertex shader.
![capture2](https://cloud.githubusercontent.com/assets/915398/17975669/9d3a55ea-6ab8-11e6-926b-17fa8af49bfa.JPG)

pjcozzi commented:
> This also lays the foundation for dynamic styling in the shader...

Very nice.

pjcozzi commented:
Do you still plan to do this in this PR?

>  * [ ] Possibly remove handleTranslucent check for pnts, should be able to support transparent and opaque together.

pjcozzi commented:
Should the arraybuffer/view memory concerns we discussed offline be addressed in this PR or another?

pjcozzi commented on Source/Scene/Model.js:
```diff
@@ -51,7 +51,7 @@ define([
         '../ThirdParty/Uri',
         '../ThirdParty/when',
-        './getModelAccessor',
+        './getBinaryAccessor',
```
Can you cherry-pick this into master?

lilleyse replied:
Ah yes, thanks for reminding me.

lilleyse commented:
> Do you still plan to do this in this PR?

> > Possibly remove handleTranslucent check for pnts, should be able to support transparent and opaque together.

I think I may pass on that for later. I'm a bit worried about the performance hit associated with doing all the pass/translucent checks per point. As of now supporting opaque and translucent together is only an issue for batched points, and maybe just setting the render state to translucent may be enough.

pjcozzi commented on Source/Scene/Points3DTileContent.js:
```diff
@@ -340,12 +423,80 @@ define([
+        var hasStyleableProperties = defined(styleableProperties);
+
+        // Use per-point normals to hide back-facing points.
+        var backFaceCulling = true;
```
Check out hidden point removal in http://www.crs4.it/vic/cgi-bin/multimedia-page.cgi?id=160

Could be a nice reading group paper when we are ready to implement it.

pjcozzi replied:
Also add a TODO here for how to expose this, e.g., will the Cesium API have a point cloud specific style, is this a tileset property, etc.

lilleyse commented:
> Should the arraybuffer/view memory concerns we discussed offline be addressed in this PR or another?

I'll handle it here.

pjcozzi commented:
> I think I may pass on that for later. I'm a bit worried about the performance hit associated with doing all the pass/translucent checks per point. As of now supporting opaque and translucent together is only an issue for batched points, and maybe just setting the render state to translucent may be enough.

Good plan.

pjcozzi commented on Source/Scene/Points3DTileContent.js:
```diff
@@ -340,12 +423,80 @@ define([
+            for (var name in styleableProperties) {
+                if (styleableProperties.hasOwnProperty(name)) {
+                    var property = styleableProperties[name];
+                    // TODO : this will not handle matrix types currently
```
Not part of this PR?

lilleyse replied:
I can probably get something working for this, but I'm not sure of the benefits. Would the styling language even support the concept of matrices?

pjcozzi replied:
It could...and it likely will at some point...but I would rather roadmap this since it is not in the most promient use cases, update https://github.com/AnalyticalGraphicsInc/3d-tiles/issues/2

lilleyse replied:
Added to roadmap.

pjcozzi commented on Source/Scene/Points3DTileContent.js:
```diff
@@ -340,12 +423,80 @@ define([
+                        vertexBuffer : vertexBuffer,
+                        componentsPerAttribute : componentCount,
+                        componentDatatype : componentDatatype,
+                        normalize : false, // TODO : support this in batch table binary spec?
```
Do you think we can go without it?  Do you have another killer use case other than unsigned byte RGBA, which is already in the feature table?

lilleyse replied:
Yeah we can go without it.

pjcozzi commented on Source/Scene/Points3DTileContent.js:
```diff
@@ -403,10 +562,21 @@ define([
             vs += '    vec3 position = a_position; \n';
         }
 
+        // TODO : only for testing
```
Keep this for now, right?

lilleyse replied:
I'm going to remove it. This was mainly for anyone reviewing the PR to see how it works.

pjcozzi replied:
Ah, look at the shader generation in the Batch Table, but I think we want to prefix these more specifically than `style_`.

pjcozzi commented:
Just those comments.  Really liking the direction here.

pjcozzi commented:
The points batched example in Sandcastle comes up blank here: http://cesium-dev.s3-website-us-east-1.amazonaws.com/cesium/batch-table-binary/Apps/Sandcastle/index.html?src=3D%20Tiles.html&label=Showcases

Does it for you?

lilleyse commented:
Updated. 

>The points batched example in Sandcastle comes up blank here: http://cesium-dev.s3-website-us-east-1.amazonaws.com/cesium/batch-table-binary/Apps/Sandcastle/index.html?src=3D%20Tiles.html&label=Showcases

This works now.

The memory issues are also addressed. `Model` doesn't hold on to the array buffer: https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Scene/Model.js#L3697.

pjcozzi commented on Source/Scene/Batched3DModel3DTileContent.js:
```diff
@@ -238,6 +238,8 @@ define([
                 // Has a batch table binary
                 batchTableBinary = new Uint8Array(arrayBuffer, byteOffset, batchTableBinaryByteLength);
+                // Copy the batchTableBinary section and let the underlying ArrayBuffer be freed
+                batchTableBinary = new Uint8Array(batchTableBinary);
```
Does any part of Model.js need to change?

lilleyse replied:
No

pjcozzi commented:
Part of #3241

pjcozzi commented on Source/Scene/Points3DTileContent.js:
```diff
@@ -340,12 +434,81 @@ define([
+
+        // TODO : How to expose this? Will this be part of the point cloud styling or a property of the tileset?
+        // Use per-point normals to hide back-facing points.
+        var backFaceCulling = false;
```
I changed this to `false` by default since many users will want go inside point clouds, e.g., like the church model.
